lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,88 +3642,1015 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+    }
+    return getBuilder().createStruct(loc, structLoweredTy,
+                                     bbDifferentialValues);
+  }
+
+  //--------------------------------------------------------------------------//
+  // Tangent materialization
+  //--------------------------------------------------------------------------//
+
+  void emitZeroIndirect(CanType type, SILValue bufferAccess,
+                        SILLocation loc) {
+    auto builder = getDifferentialBuilder();
+    auto tangentSpace = getTangentSpace(type);
+    assert(tangentSpace && "No tangent space for this type");
+    switch (tangentSpace->getKind()) {
+      case VectorSpace::Kind::Vector:
+        emitZeroIntoBuffer(builder, type, bufferAccess, loc);
+        return;
+      case VectorSpace::Kind::Tuple: {
+        auto tupleType = tangentSpace->getTuple();
+        SmallVector<SILValue, 8> zeroElements;
+        for (unsigned i : range(tupleType->getNumElements())) {
+          auto eltAddr = builder.createTupleElementAddr(loc, bufferAccess, i);
+          emitZeroIndirect(tupleType->getElementType(i)->getCanonicalType(),
+                           eltAddr, loc);
+        }
+        return;
+      }
+      case VectorSpace::Kind::Function: {
+        llvm_unreachable(
+            "Unimplemented: Emit thunks for abstracting zero initialization");
+      }
+    }
+  }
+
+  SILValue emitZeroDirect(CanType type, SILLocation loc) {
+    auto builder = getDifferentialBuilder();
+    auto silType = getModule().Types.getLoweredLoadableType(
+        type, ResilienceExpansion::Minimal);
+    auto *buffer = builder.createAllocStack(loc, silType);
+    auto *initAccess = builder.createBeginAccess(
+        loc, buffer, SILAccessKind::Init, SILAccessEnforcement::Static,
+        /*noNestedConflict*/ true, /*fromBuiltin*/ false);
+    emitZeroIndirect(type, initAccess, loc);
+    builder.createEndAccess(loc, initAccess, /*aborted*/ false);
+    auto readAccess = builder.createBeginAccess(
+        loc, buffer, SILAccessKind::Read, SILAccessEnforcement::Static,
+        /*noNestedConflict*/ true, /*fromBuiltin*/ false);
+    auto *loaded = builder.createLoad(
+        loc, readAccess, getBufferLOQ(type, getDifferential()));
+    builder.createEndAccess(loc, readAccess, /*aborted*/ false);
+    builder.createDeallocStack(loc, buffer);
+    return loaded;
+  }
+
+  SILValue materializeTangentDirect(
+      AdjointValue val, SILLocation loc) {
+    assert(val.getType().isObject());
+    LLVM_DEBUG(getADDebugStream() <<
+               "Materializing tangents for " << val << '\n');
+    assert(val.getKind() != AdjointValueKind::Aggregate
+           && "Tangent values cannot be aggregated in forward mode.");
+    switch (val.getKind()) {
+      case AdjointValueKind::Zero: {
+        auto zeroVal = emitZeroDirect(val.getSwiftType(), loc);
+        return zeroVal;
+      }
+      // Aggregate case here is necessary until we update 'AdjointValue' to be
```

Comment:
* Not sure I understand this comment.
* If an assertion is really necessary here, use `llvm_unreachable("message")` below the case label instead of calling `assert` above the switch.

Reply:
Comment was trying to explain how you should never have a adjoint value of kind `AdjointValueKind::Aggregate` in JVP generation since we never do any sort of `tan[x] += tan[y]` type of operation. We never accumulate tangent values, only set them once or override them. I'll put the llvm_unreachable in the case instead.

Reply:
Hmm, I think you misunderstood what `AdjointValueKind::Aggregate` is completely. It has nothing to do with accumulation. It's for representing tuples and structs for efficient element extraction.

Reply:
So an assertion here is definitely the wrong thing. You should create aggregate adjoint values when differentiating `tuple` and `struct` instructions and match the `Aggregate` case when differentiating `tuple_extract` and `struct_extract`.

Reply:
Ahh yes I mistook the aggregate in terms of us accumulating/aggregating tangent values in pullback emitter, and not tuples and structs which "aggregate" elements/values.